### PR TITLE
[Pending] Energy monitoring - New Add date range recalculation for energy monitoring with improved jobs and tests

### PR DESCRIPTION
**Pull request by Terdious**
### Pull Request check-list

To ensure your Pull Request can be accepted as fast as possible, make sure to review and check all of these items:

- [x] If your changes affects code, did your write the tests?
- [x] Are tests passing? (`npm test` on both front/server)
- [x] Is the linter passing? (`npm run eslint` on both front/server)
- [x] Did you run prettier? (`npm run prettier` on both front/server)
- [x] If you are adding a new features/services, did you run integration comparator? (`npm run compare-translations` on front)
- [x] Did you test this pull request in real life? With real devices? If this development is a big feature or a new service, we recommend that you provide a Docker image to the community ([forum](https://community.gladysassistant.com/)) for testing before merging.
- [ ] If your changes modify the API (REST or Node.js), did you modify the API documentation? (Documentation is based on comments in code)
- [x] If you are adding a new features/services which needs explanation, did you modify the user documentation? See [the GitHub repo](https://github.com/GladysAssistant/v4-website) and the [website](https://gladysassistant.com).
- [ ] Did you add fake requests data for the demo mode (`front/src/config/demo.js`) so that the demo website is working without a backend? (if needed) See [https://demo.gladysassistant.com](https://demo.gladysassistant.com).

NOTE: these things are not required to open a PR and can be done afterwards / while the PR is open.

### Description of change

* **Translate**
  * +126 lines / -0 lines

* **Front**
  * +792 lines / -132 lines

* **Server**
  * +773 lines / -141 lines

* **Tests**
  * +1359 lines / -129 lines

- Add start/end date support for energy consumption and cost recalculation (full or selected features).
- Purge recalculated consumption and cost states within the selected period before recomputing.
- Return `job_id` on recalculation endpoints and improve detached job handling/progress updates.
- Add `destroyStatesBetween` helper and related tests.
- Validate date inputs in energy-monitoring range controllers (`YYYY-MM-DD`).
- Restore `ENERGY_INDEX_LAST_PROCESSED` after partial/selector-based recalculations to avoid side effects on incremental runs.
- Improve Energy Monitoring UI for feature selection/reordering and date-range driven recalculation flows.

## Release Notes

* **New Features**
  * Added feature selection for energy recalculation workflows, allowing users to choose specific features for consumption and cost calculations.
  * Introduced date range picker for consumption and cost calculations, enabling recalculation for specific time periods.
  * Extended rename toggle functionality in device list drag-and-drop component.

* **Documentation**
  * Added comprehensive translations in English, French, and German for energy monitoring recalculation flows and feature selection interfaces.

* **Tests**
  * Added comprehensive test coverage for new date-range calculation features and feature selector handling.

### CODE REVIEW

File: front/src/components/device/index.js [hunk 1]
```diff
@@ -51,6 +58,88 @@ class EditDevicePage extends Component {
...
+        features: {
+          $push: [consumptionFeature, costFeature]
+        }
+      });
+    }
+
+    this.setState({
+      device: deviceUpdate
+    });
+  }
```
[the remark this answers is not in the document]
Comment (Terdious) [reply front/src/components/device/index.js]: Added the self‑reference guard in [index.js](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/home/terdieu/.vscode-server/extensions/openai.chatgpt-0.4.71-linux-x64/webview/#) by checking [defaultElectricMeterFeatureId !== parentFeature.id](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/home/terdieu/.vscode-server/extensions/openai.chatgpt-0.4.71-linux-x64/webview/#) before setting energy_parent_id, preventing local corrupted hierarchy if the API returns the current feature id.

File: server/services/energy-monitoring/lib/energy-monitoring.calculateConsumptionFromIndexFromBeginning.js [hunk 2]
```diff
@@ -14,14 +14,19 @@ const { ENERGY_INDEX_FEATURE_TYPES } = require('../utils/constants');
...
- * @description Calculate consumption from index for all 30-minute windows from the beginning of the instance until now.
+ * @description Calculate consumption from index for all 30-minute windows from the beginning.
  * This function finds the oldest device state for energy index devices and processes all 30-minute windows.
+ * @param {Array<string>} featureSelectors - Optional whitelist of consumption feature selectors.
  * @param {string} jobId - The job id.
  * @returns {Promise<null>} Return null when finished.
- * @example
- * calculateConsumptionFromIndexFromBeginning('12345678-1234-1234-1234-1234567890ab');
+ * @example <caption>Recalculate full history</caption>
+ * calculateConsumptionFromIndexFromBeginning(null, [], null, '12345678-1234-1234-1234-1234567890ab');
```
[the remark this answers is not in the document]
Comment (Terdious) [reply server/services/energy-monitoring/lib/energy-monitoring.calculateConsumptionFromIndexFromBeginning.js]: Done — fixed the JSDoc example in [energy-monitoring.calculateConsumptionFromIndexFromBeginning.js](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/home/terdieu/.vscode-server/extensions/openai.chatgpt-0.4.71-linux-x64/webview/#) to match the actual (featureSelectors, jobId) signature.

File: front/src/components/device/UpdateDeviceFeature.jsx [hunk 3]
```diff
@@ -189,6 +213,45 @@ class UpdateDeviceFeature extends Component {
...
+                  </p>
+                  <button
+                    onClick={this.createEnergyConsumptionFeatures}
+                    class="btn btn-primary btn-sm d-block w-100 text-center"
+                  >
+                    <Text id="integration.energyMonitoring.featureHint.createButton" />
+                  </button>
+                </div>
+              </div>
+            )}
```
[the remark this answers is not in the document]
Comment (Terdious) [reply front/src/components/device/UpdateDeviceFeature.jsx]: I removed the manual “create energy features” UI from the device edit flow (aligned with commit 06070057 on energy-monitoring-tasmota-and-other).
Changes:

removed the create button in [UpdateDeviceFeature.jsx](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/home/terdieu/.vscode-server/extensions/openai.chatgpt-0.4.71-linux-x64/webview/#)
removed the createEnergyConsumptionFeatures handler and related imports in [index.js](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/home/terdieu/.vscode-server/extensions/openai.chatgpt-0.4.71-linux-x64/webview/#)
kept the shortcut link to the Energy Monitoring integration when consumption/cost features exist.
So the CodeRabbit gating request is now not applicable because the button is gone.

File: server/services/energy-monitoring/lib/energy-monitoring.calculateConsumptionFromIndexFromBeginning.js [hunk 4]
```diff
@@ -40,27 +45,44 @@ async function calculateConsumptionFromIndexFromBeginning(jobId) {
...
+          f.type === DEVICE_FEATURE_TYPES.ENERGY_SENSOR.THIRTY_MINUTES_CONSUMPTION;
+        if (!isConsumption) {
+          return false;
+        }
+        if (selectorSet.size === 0) {
+          return true;
+        }
+        const featureSelector = f.selector || f.external_id || f.id;
+        return selectorSet.has(featureSelector);
+      });
```
[the remark this answers is not in the document]
Comment (Terdious) [reply server/services/energy-monitoring/lib/energy-monitoring.calculateConsumptionFromIndexFromBeginning.js:62]: I addressed the cleanup mismatch by filtering out consumption features without a selector in calculateConsumptionFromIndexFromBeginning. Since the downstream pipeline relies on selector (getDeviceFeatureStates/saveHistoricalState/destroyStatesFrom), skipping these avoids inconsistent cleanup and prevents processing non‑cleanable features.

File: server/services/energy-monitoring/lib/energy-monitoring.calculateCostFromBeginning.js [hunk 5]
```diff
@@ -2,14 +2,19 @@ const { queueWrapper } = require('../utils/queueWrapper');
 
 /**
  * @description Calculate energy monitoring cost from the beginning.
+ * @param {Array<string>} featureSelectors - Optional whitelist of cost feature selectors.
  * @param {string} jobId - The job id.
  * @returns {Promise<null>} Return when finished.
- * @example
- * calculateCostFromBeginning('12345678-1234-1234-1234-1234567890ab');
+ * @example <caption>Recalculate full history</caption>
+ * calculateCostFromBeginning(null, [], null, '12345678-1234-1234-1234-1234567890ab');
```
[the remark this answers is not in the document]
Comment (Terdious) [reply server/services/energy-monitoring/lib/energy-monitoring.calculateCostFromBeginning.js]: Fixed the JSDoc example in [energy-monitoring.calculateCostFromBeginning.js](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/home/terdieu/.vscode-server/extensions/openai.chatgpt-0.4.71-linux-x64/webview/#) to match the actual (featureSelectors, jobId) signature.

File: server/test/services/energy-monitoring/energy-monitoring.calculateConsumptionFromIndex.test.js [hunk 6]
```diff
@@ -811,5 +814,47 @@ describe('EnergyMonitoring.calculateConsumptionFromIndex', () => {
...
+      await energyMonitoring.calculateConsumptionFromIndex(testTime, undefined, 'job-progress');
+      // Two calls: one during processing, one final at 100%
+      expect(gladys.job.updateProgress.callCount).to.equal(2);
+      const firstCall = gladys.job.updateProgress.getCall(0).args;
+      expect(firstCall[0]).to.equal('job-progress');
+      expect(firstCall[1]).to.equal(100);
+      const finalCall = gladys.job.updateProgress.getCall(1).args;
+      expect(finalCall[0]).to.equal('job-progress');
+      expect(finalCall[1]).to.equal(100);
+    });
```
[the remark this answers is not in the document]
Comment (Terdious) [reply server/test/services/energy-monitoring/energy-monitoring.calculateConsumptionFromIndex.test.js:858]: Updated calls in [energy-monitoring.calculateConsumptionFromIndex.test.js](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/home/terdieu/.vscode-server/extensions/openai.chatgpt-0.4.71-linux-x64/webview/#) to use the 3‑parameter signature (passed undefined as the second argument and jobId as the third).

File: front/src/config/i18n/de.json [hunk 7]
```diff
@@ -689,13 +689,53 @@
...
+      "calculateSelectedFeaturesDescription": "Wählen Sie eine oder mehrere Energiefunktionen, um deren Verbrauch und Kosten von Anfang an neu zu berechnen.",
+      "selectFeatures": "Funktionen auswählen",
+      "selectFeaturesPlaceholder": "Funktionen auswählen oder leer lassen",
+      "noFeaturesAvailable": "Keine passenden Funktionen gefunden.",
+      "calculateSelectedFeaturesFromBeginning": "Verbrauch und Kosten der ausgewählten Funktionen von Anfang an berechnen",
+      "calculateSelectedFeaturesOnPeriod": "Verbrauch und Kosten der ausgewählten Funktionen für den Zeitraum berechnen",
+      "calculatingSelectedFeaturesFromBeginning": "Berechnung von Verbrauch und Kosten der ausgewählten Funktionen wird gestartet...",
+      "calculateSelectedFeaturesCostFromBeginning": "Kosten der ausgewählten Funktionen von Anfang an berechnen",
+      "calculateSelectedFeaturesCostOnPeriod": "Kosten der ausgewählten Funktionen für den Zeitraum berechnen",
+      "calculatingSelectedFeaturesCostFromBeginning": "Kostenberechnung für ausgewählte Funktionen wird gestartet...",
```
[the remark this answers is not in the document]
Comment (Terdious) [reply front/src/config/i18n/de.json]: Aligned German in‑progress labels to use “läuft…” in [de.json](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/home/terdieu/.vscode-server/extensions/openai.chatgpt-0.4.71-linux-x64/webview/#).

File: front/src/config/i18n/de.json [hunk 8]
```diff
@@ -727,7 +767,15 @@
...
-      "fixing": "Behebung läuft..."
+      "fixing": "Behebung läuft...",
+      "featureHint": {
+        "title": "Energieüberwachung",
+        "description": "Diese Funktion kann zur Energieüberwachung verwendet werden. Klicken Sie auf die Schaltfläche unten, um automatisch die Verbrauchs- und Kostenverfolgungsfunktionen zu erstellen.",
+        "createButton": "Energieüberwachungsfunktionen erstellen",
+        "hierarchy": "Ordne sie der Hierarchie „Energieüberwachung“ zu (Tab „Meine Geräte“), um Kosten zu verknüpfen.",
+        "recalc": "Dann Verbrauch/Kosten im Tab „Einstellungen“ neu berechnen.",
+        "goToIntegration": "Energieüberwachung öffnen"
+      }
```
[the remark this answers is not in the document]
Comment (Terdious) [reply front/src/config/i18n/de.json]: Switched the German featureHint lines to formal tone in [de.json](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/home/terdieu/.vscode-server/extensions/openai.chatgpt-0.4.71-linux-x64/webview/#).

File: front/src/config/i18n/en.json [hunk 9]
```diff
@@ -1783,7 +1818,15 @@
...
-      "fixing": "Fixing..."
+      "fixing": "Fixing...",
+      "featureHint": {
+        "title": "Energy monitoring",
+        "description": "This feature can be used for energy monitoring. Click the button below to automatically create the consumption and cost tracking features.",
+        "createButton": "Create energy monitoring features",
+        "hierarchy": "Place it in the “Energy monitoring” hierarchy (My devices tab) to link cost features.",
+        "recalc": "Then rerun consumption/cost from the Parameters tab.",
+        "goToIntegration": "Open Energy monitoring"
+      }
```
[the remark this answers is not in the document]
Comment (Terdious) [reply front/src/config/i18n/en.json]: Already fixed. The featureHint.recalc string now references “Settings tab” in [en.json](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/home/terdieu/.vscode-server/extensions/openai.chatgpt-0.4.71-linux-x64/webview/#), so no further change needed.

File: server/test/utils/duckdb.js [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+const clearDuckDb = async () => {
+  // Delete children first to avoid FK issues.
+  // eslint-disable-next-line no-restricted-syntax
+  for (const table of DEFAULT_TABLES) {
+    try {
+      // eslint-disable-next-line no-await-in-loop
+      await db.duckDbWriteConnectionAllAsync(`DELETE FROM ${table}`);
+    } catch (e) {
+      // ignore missing tables for DuckDB
+    }
```
[the remark this answers is not in the document]
Comment (Terdious) [reply server/test/utils/duckdb.js:31]: In [duckdb.js](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/home/terdieu/.vscode-server/extensions/openai.chatgpt-0.4.71-linux-x64/webview/#), only missing-table errors are ignored; unexpected cleanup errors now rethrow.

[the remark this answers is not in the document]
Comment (Terdious) [reply server/test/utils/duckdb.js:31]: In [duckdb.js](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/home/terdieu/.vscode-server/extensions/openai.chatgpt-0.4.71-linux-x64/webview/#), only missing-table errors are ignored; unexpected cleanup errors now rethrow.

File: front/src/components/device/UpdateDeviceFeature.jsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (Terdious) [reply front/src/components/device/UpdateDeviceFeature.jsx]: This comment is now outdated. The manual “create energy features” UI has been removed from [UpdateDeviceFeature.jsx](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/home/terdieu/.vscode-server/extensions/openai.chatgpt-0.4.71-linux-x64/webview/#), and the handler/binding has been removed from [index.js](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/home/terdieu/.vscode-server/extensions/openai.chatgpt-0.4.71-linux-x64/webview/#). There is no showCreateEnergyFeaturesButton or createEnergyConsumptionFeatures left. Once the updated commits are pushed, this suggestion will no longer apply.

File: server/test/services/energy-monitoring/energy-monitoring.calculateConsumptionFromIndex.test.js [hunk 11]
```diff
@@ -811,5 +814,47 @@ describe('EnergyMonitoring.calculateConsumptionFromIndex', () => {
...
+
+      device.getDeviceFeatureStates = fake((selector) => {
+        if (selector === 'test-energy-device-index') {
+          return [
+            { created_at: '2023-10-03T13:30:00.000Z', value: 1000 },
+            { created_at: '2023-10-03T14:00:00.000Z', value: 1010 },
+          ];
+        }
+        return [];
+      });
```
[the remark this answers is not in the document]
Comment (Terdious) [reply server/test/services/energy-monitoring/energy-monitoring.calculateConsumptionFromIndex.test.js:832]: Fixed the selector mismatch in [energy-monitoring.calculateConsumptionFromIndex.test.js](https://vscode-remote+wsl-002bubuntu.vscode-resource.vscode-cdn.net/home/terdieu/.vscode-server/extensions/openai.chatgpt-0.4.71-linux-x64/webview/#) to use 'test-device-index'.

File: server/test/services/energy-monitoring/energy-monitoring.calculateCostFrom.test.js [hunk 12]
```diff
@@ -666,4 +673,180 @@ describe('EnergyMonitoring.calculateCostFrom', () => {
...
+      .stub(gladys.device.energySensorManager, 'getRootElectricMeterDevice')
+      .returns({ id: 'meter-feature', device_id: 'meter-device' });
+    const destroyBetweenStub = sinon.stub(gladys.device, 'destroyStatesBetween').resolves();
+    const priceStub = sinon
+      .stub(gladys.energyPrice, 'get')
+      .resolves([{ price_type: ENERGY_PRICE_TYPES.CONSUMPTION, price: 0.2, start_date: '2025-10-01' }]);
+    const statesStub = sinon
+      .stub(gladys.device, 'getDeviceFeatureStates')
+      .resolves([{ created_at: new Date('2025-10-01T00:00:00.000Z'), value: 1 }]);
+    const saveMultipleStub = sinon.stub(gladys.device, 'saveMultipleHistoricalStates').rejects(new Error('boom')); // triggers catch
```
[the remark this answers is not in the document]
Comment (Terdious) [reply server/test/services/energy-monitoring/energy-monitoring.calculateCostFrom.test.js]: Updated the stubbed price to 2000 (integer / 4 decimals) for consistency.

File: server/test/services/energy-monitoring/energy-monitoring.calculateCostFrom.test.js [hunk 13]
```diff
@@ -666,4 +673,180 @@ describe('EnergyMonitoring.calculateCostFrom', () => {
...
+        null,
+        '2025-10-02',
+      );
+      expect(res).to.equal(null);
+      expect(destroyBetweenStub.calledOnce).to.equal(true);
+    } finally {
+      priceStub.restore();
+      destroyBetweenStub.restore();
+    }
+  });
```
[the remark this answers is not in the document]
Comment (Terdious) [reply server/test/services/energy-monitoring/energy-monitoring.calculateCostFrom.test.js]: I documented the polymorphic signature in calculateCostFrom JSDoc.
I did not change the “invalid start date type” test because it legitimately verifies the guard (invalid input → null). Let me know if you want it changed anyway.

**Comment by Terdious** [comment]
@Pierre-Gilles ,
Depuis peu, le cypress run casse souvent, obligé de relancer les checks souvent et ca casse toujours au niveau des scenes au même endroit. Généralement après relaunch plus de soucis. Normal à ton avis ?
image

image

**Comment by Terdious** [comment]
@coderabbitai full review

Please review the full PR again, but keep feedback limited to the actual scope of this PR and the files impacted by it. Focus on concrete regressions, inconsistencies with existing Gladys patterns, and real risks for current behavior. Avoid speculative improvements, overengineering, or suggestions unrelated to the current implementation goals.

Please review this PR with these priorities:
- Compare the code style and patterns to the existing Gladys codebase and avoid suggestions that diverge from current project habits.
- Focus on consistency with the current repository conventions rather than proposing new architectural patterns.
- Avoid overengineering and speculative hardening for hypothetical future use cases.
- Prioritize concrete risks: regressions, broken existing behavior, inconsistent tests, edge cases already relevant to the current implementation, and mismatch with surrounding code.
- Be pragmatic: only suggest additional guards or abstractions if they are clearly needed by the current use case.

**Comment by Terdious** [comment]
@coderabbitai status

File: server/test/services/energy-monitoring/energy-monitoring.calculateConsumptionFromIndexFromBeginning.test.js [hunk 14]
```diff
@@ -358,4 +488,105 @@ describe('EnergyMonitoring.calculateConsumptionFromIndexFromBeginning', () => {
...
+    // calc called for windows
+    expect(calcStub.called).to.equal(true);
+    // destroyStatesBetween should not throw even with missing selector feature
+    expect(destroyBetweenStub.called || destroyFromStub.called).to.equal(true);
+    calcStub.restore();
+    destroyFromStub.restore();
+    destroyBetweenStub.restore();
+    oldestStub.restore();
+    getStub.restore();
+  });
```
[the remark this answers is not in the document]
Comment (Terdious) [reply server/test/services/energy-monitoring/energy-monitoring.calculateConsumptionFromIndexFromBeginning.test.js]: Fixed. I renamed the duplicated test title to make test output unambiguous.

File: server/services/energy-monitoring/lib/energy-monitoring.calculateConsumptionFromIndexRange.js [hunk 15]
```diff
@@ -0,0 +1,278 @@
...
+    const shouldRestoreLastProcessed = Boolean((parsedEndAt && parsedEndAt < nowDate) || selectorSet.size > 0);
+    const originalLastProcessedByDeviceId = new Map();
+    if (shouldRestoreLastProcessed) {
+      devicesWithBothFeatures.forEach(({ device }) => {
+        const lastProcessedParam = device.params.find((p) => p.name === ENERGY_INDEX_LAST_PROCESSED);
+        if (lastProcessedParam && lastProcessedParam.value) {
+          originalLastProcessedByDeviceId.set(device.id, lastProcessedParam.value);
+        }
+      });
+    }
```
[the remark this answers is not in the document]
Comment (Terdious) [reply server/services/energy-monitoring/lib/energy-monitoring.calculateConsumptionFromIndexRange.js]: Fixed. In range recalculation, cursor restoration is now also triggered when failed windows are detected, even when the previous context would not restore it.

File: server/services/energy-monitoring/lib/energy-monitoring.calculateConsumptionFromIndexFromBeginning.js [hunk 16]
```diff
@@ -40,26 +45,54 @@ async function calculateConsumptionFromIndexFromBeginning(jobId) {
...
+    const shouldRestoreLastProcessed = selectorSet.size > 0;
+    const originalLastProcessedByDeviceId = new Map();
+    if (shouldRestoreLastProcessed) {
+      devicesWithBothFeatures.forEach(({ device }) => {
+        const lastProcessedParam = device.params.find((p) => p.name === ENERGY_INDEX_LAST_PROCESSED);
+        if (lastProcessedParam && lastProcessedParam.value) {
+          originalLastProcessedByDeviceId.set(device.id, lastProcessedParam.value);
+        }
+      });
+    }
```
[the remark this answers is not in the document]
Comment (Terdious) [reply server/services/energy-monitoring/lib/energy-monitoring.calculateConsumptionFromIndexFromBeginning.js]: Fixed. On from-beginning recalculation, if any window fails, the original ENERGY_INDEX_LAST_PROCESSED value is now restored to avoid silently advancing the cursor.

File: server/services/energy-monitoring/api/energy-monitoring.controller.js [hunk 17]
```diff
@@ -7,9 +43,11 @@ module.exports = function EnergyMonitoringController(energyMonitoringHandler) {
    * @apiGroup EnergyMonitoring
    */
   async function calculateCostFromBeginning(req, res) {
-    energyMonitoringHandler.calculateCostFromBeginning();
+    const featureSelectors = Array.isArray(req.body && req.body.feature_selectors) ? req.body.feature_selectors : [];
+    const job = await energyMonitoringHandler.calculateCostFromBeginning(featureSelectors);
```
[the remark this answers is not in the document]
Comment (Terdious) [reply server/services/energy-monitoring/api/energy-monitoring.controller.js:70]: Fixed. I added strict request validation for feature_selectors (must be an array of non-empty strings), so malformed payloads now return BadParameters instead of widening to full recalculation.

File: front/src/routes/integration/all/energy-monitoring/EnergyMonitoring.jsx [hunk 18]
```diff
@@ -132,38 +167,307 @@ class EnergyMonitoringPage extends Component {
...
+
+  getSelectedRecalculationSelectors = () => {
+    const consumptionSet = new Set();
+    const costSet = new Set();
+    (this.state.selectedFeaturesForRecalc || []).forEach(opt => {
+      if (opt && opt.consumptionSelector) consumptionSet.add(opt.consumptionSelector);
+      if (opt && opt.costSelector) costSet.add(opt.costSelector);
+    });
+    return { consumptionSelectors: Array.from(consumptionSet), costSelectors: Array.from(costSet) };
+  };
```
[the remark this answers is not in the document]
Comment (Terdious) [reply front/src/routes/integration/all/energy-monitoring/EnergyMonitoring.jsx:226]: I fixed this by aggregating all descendant selectors for a selected root (both DAILY_* and THIRTY_MINUTES_* branches), instead of keeping only the first match.

File: front/src/config/i18n/fr.json [hunk 19]
```diff
@@ -587,13 +587,53 @@
...
       "calculateCostFromBeginningDescription": "Ceci calculera le coût de votre consommation d'énergie en fonction de vos données de consommation et des tarifs d'électricité. Utilisez ceci après avoir des données de consommation et configuré vos tarifs d'électricité.",
       "calculateCostFromBeginning": "Calculer le coût depuis le début",
+      "calculateCostOnPeriod": "Calculer le coût sur la période",
+      "calculatingCostFromBeginning": "Lancement du calcul du coût...",
+      "calculateCostFromBeginningStarted": "Le calcul du coût a été lancé. Vous pouvez suivre le progrès dans les tâches en arrière-plan.",
+      "calculateCostFromBeginningError": "Une erreur est survenue lors du lancement du calcul du coût.",
+      "calculateConsumptionAndCostFromBeginning": "Calculer la consommation et le coût depuis le début",
+      "calculateConsumptionAndCostOnPeriod": "Calculer la consommation et le coût sur la période",
+      "calculateSelectedFeaturesDescription": "Sélectionnez une ou plusieurs fonctionnalités énergie pour recalculer leur consommation et leur coût depuis le début.",
```
[the remark this answers is not in the document]
Comment (Terdious) [reply front/src/config/i18n/fr.json]: Fixed. I updated the selected-features description to be period-agnostic.

File: front/src/config/i18n/en.json [hunk 20]
```diff
@@ -1758,13 +1758,53 @@
...
       "calculateConsumptionFromBeginningError": "An error occurred while starting the consumption calculation.",
       "calculateCostFromBeginningDescription": "This will calculate the cost of your energy consumption based on your consumption data and electricity prices. Use this after you have consumption data and configured your electricity prices.",
       "calculateCostFromBeginning": "Calculate cost from the beginning",
+      "calculateCostOnPeriod": "Calculate cost for the period",
+      "calculatingCostFromBeginning": "Starting cost calculation...",
+      "calculateCostFromBeginningStarted": "Cost calculation started. You can follow the progress in background tasks.",
+      "calculateCostFromBeginningError": "An error occurred while starting the cost calculation.",
+      "calculateConsumptionAndCostFromBeginning": "Calculate consumption and cost from the beginning",
+      "calculateConsumptionAndCostOnPeriod": "Calculate consumption and cost for the period",
+      "calculateSelectedFeaturesDescription": "Select one or more energy features to recalculate their consumption and cost from the beginning.",
```
[the remark this answers is not in the document]
Comment (Terdious) [reply front/src/config/i18n/en.json]: Fixed. I updated the selected-features description to be period-agnostic.

File: front/src/routes/integration/all/energy-monitoring/EnergyMonitoring.jsx [hunk 21]
```diff
@@ -132,38 +167,307 @@ class EnergyMonitoringPage extends Component {
...
+        feature_selectors: consumptionSelectors,
+        ...this.getRecalculateDatePayload()
+      });
+      if (!consumptionResponse || consumptionResponse.success !== true || !consumptionResponse.job_id) {
+        throw new Error('job_not_created');
+      }
+      const costResponse = await this.props.httpClient.post(costUrl, {
+        feature_selectors: costSelectors,
+        ...this.getRecalculateDatePayload()
+      });
```
[the remark this answers is not in the document]
Comment (Terdious) [reply front/src/routes/integration/all/energy-monitoring/EnergyMonitoring.jsx:457]: We are intentionally keeping the current 2-job flow in this PR scope.

The consumption and cost jobs are separate by design, and execution is serialized by the energy-monitoring queue, so cost does not run in parallel before consumption completes.
We agree there is UI-level non-atomicity between the two POST calls, but changing this now would significantly increase UI/UX and backend orchestration complexity for this PR.

We already have a follow-up workstream for this area: [WIP] Add detailed task/job logging for energy recalculation (#2412).
That work will provide precise task progress/status (including crash/restart scenarios), so users can rerun only cost on already-processed consumption data, or rerun consumption+cost only on the remaining time range.

For this PR, we keep the current behavior.

**Comment by Terdious** [comment]
@Pierre-Gilles ,  should I close the PR now?
